server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -963,11 +963,12 @@ protected final IndexingStrategy planIndexingAsNonPrimary(Index index) throws IO
                 assert maxSeqNoOfNonAppendOnlyOperations.get() >= index.seqNo() : "max_seqno of non-append-only was not updated;" +
                     "max_seqno non-append-only [" + maxSeqNoOfNonAppendOnlyOperations.get() + "], seqno of index [" + index.seqNo() + "]";
             }
-            versionMap.enforceSafeAccess();
             // unlike the primary, replicas don't really care to about creation status of documents
             // this allows to ignore the case where a document was found in the live version maps in
             // a delete state and return false for the created flag in favor of code simplicity
-            if (index.seqNo() <= localCheckpointTracker.getProcessedCheckpoint()){
+            final long localCheckpoint = localCheckpointTracker.getProcessedCheckpoint();
+            final long maxSeqNoOfUpdatesOrDeletes = getMaxSeqNoOfUpdatesOrDeletes();
+            if (index.seqNo() <= localCheckpoint) {
```

Comment:
I will work on a follow-up to replace `index.seqNo() <= localCheckpoint` with `hasBeenProcessedBefore(index)`.

Reply:
I think we should do it as part of the PR. This will simplify the logic in InternalEngine I think.

Reply:
I planned to apply this change to also deletes and no-ops so we can always have at most one document (with a valid primary term) for each sequence number in Lucene.

Reply:
I applied this change for indexing operations in this PR.